### PR TITLE
Fix ConsoleInput::read() exception when false was returned by readline() or fgets()

### LINKED ISSUE #14350: ConsoleInput->read() throws Exception at stdin end
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 4.0.4
* Platform and Target: Ubuntu 18.04.4, nginx/1.14.0

### What you did
I run something like `cat file.txt | cake command`
Inside the command it's about the following code:
```
$ConsoleInput = new ConsoleInput();
while ($ConsoleInput->dataAvailable()) {
    $result = $ConsoleInput->read();
    if ($result !== false) {
        $this->myMethod($result)
    }
}
```

### What happened
At the end of the input stream (after the last row of my file) I get an exception from the read method:
```
Exception: strlen() expects parameter 1 to be string, boolean given
In [/opt/openitc/frontend/vendor/cakephp/cakephp/src/Console/ConsoleInput.php, line 64]

2020-03-10 17:10:19 Error: [TypeError] strlen() expects parameter 1 to be string, boolean given in /opt/openitc/frontend/vendor/cakephp/cakephp/src/Console/ConsoleInput.php on line 64                                                                                                               
Stack Trace:                                                                                                                                       
- /opt/openitc/frontend/vendor/cakephp/cakephp/src/Console/ConsoleInput.php:64                                                                     
- /opt/openitc/frontend/src/Lib/MyCustomLib.php:581                                                                                      
- /opt/openitc/frontend/src/Command/MyCommand.php:100                                                                                           
- /opt/openitc/frontend/vendor/cakephp/cakephp/src/Console/BaseCommand.php:175                                                                     
- /opt/openitc/frontend/vendor/cakephp/cakephp/src/Console/CommandRunner.php:336                                                                   
- /opt/openitc/frontend/vendor/cakephp/cakephp/src/Console/CommandRunner.php:171                                                                   
- /opt/openitc/frontend/bin/cake.php:12
```

### What you expected to happen
I expect that there will be no more output and no errors.

### PR DESCRIPTION
**Pull request by victoreassis**
Solves #14350 

Make `ConsoleInput::read()` return null when false is returned by the calls to ```readline()``` or ```fgets()```.

I'm not sure how to test this, btw.

### CODE REVIEW

File: src/Console/ConsoleInput.php [hunk 1]
```diff
@@ -55,20 +55,25 @@ public function __construct(string $handle = 'php://stdin')
...
+     * @return string|null The value of the stream. Null on EOF.
      */
-    public function read(): string
+    public function read(): ?string
     {
         if ($this->_canReadline) {
             $line = readline('');
-            if (strlen($line) > 0) {
+
+            if (is_string($line) && strlen($line) > 0) {
```
Comment (ADmad) [inline src/Console/ConsoleInput.php]:
```suggestion
            if ($line !== false && strlen($line) > 0) {
```

File: src/Console/ConsoleInput.php [hunk 2]
```diff
@@ -55,20 +55,25 @@ public function __construct(string $handle = 'php://stdin')
...
      * Read a value from the stream
      *
-     * @return string The value of the stream
+     * @return string|null The value of the stream. Null on EOF.
      */
-    public function read(): string
+    public function read(): ?string
     {
         if ($this->_canReadline) {
             $line = readline('');
```
Comment (ADmad) [inline src/Console/ConsoleInput.php]: PHP manual claims `readline()` always returns string https://www.php.net/manual/en/function.readline.php :angry:.

File: src/Console/ConsoleInput.php (hunk 1 above)
Comment (victoreassis) [reply src/Console/ConsoleInput.php]: Will do that.

File: src/Console/ConsoleInput.php (hunk 2 above)
Comment (victoreassis) [reply src/Console/ConsoleInput.php]: That's right. The only reference I could find was this note: https://www.php.net/manual/en/function.readline.php#86719
Both readline and fgets claims to return only string.

File: src/Console/ConsoleIo.php [hunk 3]
```diff
@@ -489,7 +489,7 @@ protected function _getInput(string $prompt, ?string $options, ?string $default)
         $this->_out->write('<question>' . $prompt . "</question>$optionsText\n$defaultText> ", 0);
         $result = $this->_in->read();
 
-        $result = trim($result);
+        $result = $result !== null ? trim($result) : '';
```
Comment (ADmad) [inline src/Console/ConsoleIo.php]:
```suggestion
        $result = $result === null ?  '' : trim($result);
```

Comment (victoreassis) [reply src/Console/ConsoleIo.php]: There's a difference between these two?

Comment (ADmad) [reply src/Console/ConsoleIo.php]: You avoid an unnecessary negation and cognitive load.

**Comment by othercorey** [comment]
Can you open a php issue for the missing documentation?

File: src/Console/ConsoleIo.php [hunk 4]
```diff
@@ -489,7 +489,7 @@ protected function _getInput(string $prompt, ?string $options, ?string $default)
         $this->_out->write('<question>' . $prompt . "</question>$optionsText\n$defaultText> ", 0);
         $result = $this->_in->read();
 
-        $result = trim($result);
+        $result = $result === null ?  '' : trim($result);
```
Comment (markstory) [inline src/Console/ConsoleIo.php]:
```suggestion
        $result = $result === null ? '' : trim($result);
```

**Comment by markstory** [comment]
Thank you 👍

**Comment by othercorey** [comment]
@victoreassis Can you link the php error report for the readline documentation if you create it?

**Comment by victoreassis** [comment]
> @victoreassis Can you link the php error report for the readline documentation if you create it?

Sorry, I had forgotten about that. Done: https://bugs.php.net/bug.php?id=79388

**Comment by othercorey** [comment]
Perfect, thanks!

**Pull request merged**